### PR TITLE
Update Dependencies for Enhanced Compatibility with Newer Python Versions

### PR DESCRIPTION
**Pull request by vincentadam87**
## Configuration changes in `pyproject.toml`
1. **cmake**: `">=3.18.0,<3.19.0"` from `"~3.18.0"`
2. **importlib-metadata**: `">=4.4,<5.0"` from `"importlib_metadata", "^1.6"`
3. **numpy**: `">=1.18.0,<2.0.0"` from `"^1.18.0"`
4. **python**: `">=3.7,<3.12"` from `"~3.7"`
5. **tensorflow**:
   - **Python 3.7 to 3.8**: `">=2.4.0,<2.7.0"`
   - **Python 3.9**: `">=2.5.0,<2.7.0"`
   - **Python 3.10**: `">=2.7.0,<2.9.0"`
   - **Python 3.11**: `">=2.8.0,<2.16.0"`
6. **mypy**: 
   -  **Python 3.7 to 3.8**: `">=1.1.0,<1.3.0"`
   -  **Python 3.9**: `">=1.1.0,<1.3.0"`
   -  **Python 3.10**: `">=1.1.0,<1.3.0"`
    - **Python 3.11**: `"1.8.0"`
7. **pytest**: Updated to `"6.2.5"` from `"^5.3.5"`
8. **pytest-black**: Updated to `">=0.3.8"` from `"^0.3.8"`

## Gihub workflows changes
1. Expanded Python Versions: Added Python 3.8.12, 3.9.12, and 3.10.4 to the matrix.
2. Updated Poetry Version: Upgraded from 1.1.6 to 1.1.12.
3. Updated Build Tools: Installed latest build tools including g++-14.
4. New C++ Build Steps: Added steps for configuring and building a C++ library with CMake.

### CODE REVIEW

File: .github/workflows/quality-check.yaml [hunk 1]
```diff
@@ -20,34 +20,56 @@ on:
...
-          sudo apt-get install cmake g++-11
-      # ------------ build and install package
+          sudo apt-get update
+          sudo apt-get install -y build-essential cmake g++-14
+
+      - name: Configure and Build C++ Library
+        run: |
+          mkdir -p banded_matrices/build
+          cd banded_matrices/build
+          cmake .. -DCMAKE_BUILD_TYPE=Release
```
Comment (johnamcleod) [inline .github/workflows/quality-check.yaml:63]: I would be tempted to use the `-B` option of `cmake`: https://cmake.org/cmake/help/latest/manual/cmake.1.html#generate-a-project-buildsystem

File: .github/workflows/quality-check.yaml [hunk 2]
```diff
@@ -20,34 +20,56 @@ on:
...
+      - name: Install Python dependencies
+        run: poetry install
+
+      - name: Configure environment variables for CMake
+        run: |
+          echo "PYTHON_BIN=$(which python)" >> $GITHUB_ENV
+
+      - name: Verify TensorFlow installation
+        run: |
+          python -c "import tensorflow as tf; print(tf.__version__)"
```
Comment (johnamcleod) [inline .github/workflows/quality-check.yaml:52]: Was this added for debugging? Could it be removed now?

File: .github/workflows/quality-check.yaml [hunk 3]
```diff
@@ -20,34 +20,56 @@ on:
...
           pip install -U pip poetry twine
           poetry config virtualenvs.create false
-      # ------------ install tools
+
+      - name: Install Python dependencies
+        run: poetry install
+
+      - name: Configure environment variables for CMake
+        run: |
+          echo "PYTHON_BIN=$(which python)" >> $GITHUB_ENV
```
Comment (johnamcleod) [inline .github/workflows/quality-check.yaml:48]: It might be better to define the environment variable in a build step specifically: https://docs.github.com/en/actions/learn-github-actions/variables#defining-environment-variables-for-a-single-workflow

File: pyproject.toml [hunk 4]
```diff
@@ -22,18 +22,18 @@ build = "build.py"
 [tool.poetry.dependencies]
 cmake = "~3.18.0"
 importlib-metadata = ">=4.4,<5.0"
-numpy = "^1.18.0"
-python = ">=3.7,<3.9"
-tensorflow = "~2.4.0"
+numpy = "^1.21.0"
+python = ">=3.10,<3.11"
+tensorflow = "~2.8.0"
```
Comment (johnamcleod) [inline pyproject.toml:27]: This is a very old version of `tensorflow`. Now the build is running it would be worth investigating how close to the current version we can get.

File: pyproject.toml [hunk 5]
```diff
@@ -51,7 +60,7 @@ test = "pytest --pylint --mypy --black --isort --cache-clear -ra -v --cov banded
 black = "black ."
 isort = "isort --atomic -y"
 format = "task isort && task black"
-check_format = "pytest -v --cache-clear --black --isort -m black,isort"
+check_format = "pytest -v --cache-clear --black --isort -m black_isort"
```
Comment (johnamcleod) [inline pyproject.toml:63]: What is the reason for this change?

File: setup.py
Comment (johnamcleod) [inline setup.py:1]: I am fairly confident this file can be deleted. Poetry can generate a `setup.py` file if one is needed.

**Review by johnamcleod — APPROVED** [review]
LGTM!

File: .github/workflows/quality-check.yaml (hunk 3 above)
Comment (vincentadam87) [reply .github/workflows/quality-check.yaml:48]: not sure what you mean here!

Comment (johnamcleod) [reply .github/workflows/quality-check.yaml:48]: Rather than having a separate step to define a new environment variable and piping it into the global environment, the link describes how to add an `env` statement to the step in the github action definition for defining contextual environment variables.

**Comment by vincentadam87** [comment]
Regarding the workflow, I have a question
A lot of the novelty in this PR is about doing some `mkdir` and `cd` at the right place so that the compiled files may be found.
This complexifies the workflows.
Is there a way to edit the `build.py` so that this is done there?

File: pyproject.toml (hunk 5 above)
Comment (vincentadam87) [reply pyproject.toml]: fixed

File: setup.py
Comment (vincentadam87) [reply setup.py:1]: removed

File: pyproject.toml (hunk 4 above)
Comment (vincentadam87) [reply pyproject.toml]: I have played further with higher versions
